### PR TITLE
Wasi support

### PR DESCRIPTION
**Pull request by Jacarte**
The main idea is to provide swam with a witx parser. With this parser, we can create the boilerplate project to implement WASI Api in scala in an automated way.

Check classes [ModuleParser](https://github.com/satabin/swam/compare/master...Jacarte:witx-generator?expand=1#diff-b5d9d718ac6f9e078c429a204963bf10)   and [TypesParser](https://github.com/satabin/swam/compare/master...Jacarte:witx-generator?expand=1#diff-1e93ddb43888945e52cdb728fdc63b76)

Stages:
 - [x] swam-generator
 - [x] WITX parser
 - [x] Generate WASI API interfaces from the WITX parsing
 - [x] WASI implementation

### CODE REVIEW

File: wasi/src/Module.scala [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+  var mem: Memory[IO] = null
+
+  val name = "wasi_snapshot_preview1"
+
+  def tryToExecute(a: => errnoEnum.Value) = {
+    try a.id
+    catch {
+      case x: WASIException => x.errno.id
+    }
+  }
```
Comment (satabin) [inline wasi/src/Module.scala:25]: The all point of having the `F` is to not use try/catch, but to handle effects and exceptions in the context of `F`. This all file is to javascript like, and I would love to have a pure implementation of it. It is a lot, I will probably check it out and rewrite it as it involves some advanced cats-effect elements.

**Comment by satabin** [comment]
@Jacarte what do you think about splitting this PR in two: one for witx, then one for wasi?

**Comment by Jacarte** [comment]
> @Jacarte what do you think about splitting this PR in two: one for witx, then one for wasi?

It is ok for me, on it. I just wanted to have a functional prototype first, to improve it after. Thanks for the suggestions, useful as usual :).

**Comment by satabin** [comment]
Hi @Jacarte 
So I revived my WASI branch that I started about a year ago, this is the kind of implementation I am aiming at: https://github.com/satabin/swam/blob/wasi/core/wasi/core/src/swam/runtime/wasi/Core.scala I will continue adding functions based on your implementation, but in the cats way.

**Comment by Jacarte** [comment]
Hi @satabin 

I will open two different PR then, one with the witx feature and another with the WASI implementation. I will close this one. I will migrate the WASI implementation to your branch :)

**Pull request closed without merging**